protocol/x/clob/e2e/withdrawal_gating_test.go

```diff
@@ -0,0 +1,224 @@
+					&genesis,
+					func(genesisState *feetiertypes.GenesisState) {
+						genesisState.Params = constants.PerpetualFeeParamsNoFee
+					},
+				)
+				return genesis
+			}).Build()
+
+			ctx := tApp.AdvanceToBlock(2, testapp.AdvanceToBlockOptions{})
+
+			// Create all existing orders.
+			existingOrderMsgs := make([]clobtypes.MsgPlaceOrder, len(tc.placedMatchableOrders))
+			for i, matchableOrder := range tc.placedMatchableOrders {
+				existingOrderMsgs[i] = clobtypes.MsgPlaceOrder{Order: matchableOrder.MustGetOrder()}
+			}
+			for _, checkTx := range testapp.MustMakeCheckTxsWithClobMsg(ctx, tApp.App, existingOrderMsgs...) {
+				resp := tApp.CheckTx(checkTx)
+				require.Conditionf(t, resp.IsOK, "Expected CheckTx to succeed. Response: %+v", resp)
+			}
+
+			_, err := tApp.App.Server.LiquidateSubaccounts(ctx, &api.LiquidateSubaccountsRequest{
+				LiquidatableSubaccountIds:  tc.liquidatableSubaccountIds,
+				NegativeTncSubaccountIds:   tc.negativeTncSubaccountIds,
+				SubaccountOpenPositionInfo: clobtest.GetOpenPositionsFromSubaccounts(tc.subaccounts),
+			})
+			require.NoError(t, err)
+
+			// Verify test expectations.
+			ctx = tApp.AdvanceToBlock(4, testapp.AdvanceToBlockOptions{})
+			for _, expectedSubaccount := range tc.expectedSubaccounts {
+				require.Equal(
+					t,
+					expectedSubaccount,
+					tApp.App.SubaccountsKeeper.GetSubaccount(ctx, *expectedSubaccount.Id),
+				)
+			}
+			negativeTncSubaccountSeenAtBlock, exists := tApp.App.SubaccountsKeeper.GetNegativeTncSubaccountSeenAtBlock(ctx)
+			require.Equal(t, tc.expectedWithdrawalsGated, exists)
+			require.Equal(t, tc.expectedNegativeTncSubaccountSeenAtBlock, negativeTncSubaccountSeenAtBlock)
+
+			// Verify withdrawals are blocked by trying to create a transfer message that withdraws funds.
+			withdrawMsg := sendingtypes.MsgWithdrawFromSubaccount{
+				Sender:    constants.Carl_Num0,
+				Recipient: constants.Carl_Num0.Owner,
+				AssetId:   constants.Usdc.Id,
+				Quantums:  1,
+			}
+			for _, checkTx := range testapp.MustMakeCheckTxsWithSdkMsg(
+				ctx,
+				tApp.App,
+				testapp.MustMakeCheckTxOptions{
+					AccAddressForSigning: constants.Carl_Num0.Owner,
+					Gas:                  1000000,
+					FeeAmt:               constants.TestFeeCoins_5Cents,
+				},
+				&withdrawMsg,
+			) {
+				resp := tApp.CheckTx(checkTx)
+				require.Conditionf(t, resp.IsOK, "Expected CheckTx to succeed. Response: %+v", resp)
+			}
+			ctx = tApp.AdvanceToBlock(
+				5,
+				testapp.AdvanceToBlockOptions{
+					ValidateFinalizeBlock: func(
+						ctx sdktypes.Context,
+						request abcitypes.RequestFinalizeBlock,
+						response abcitypes.ResponseFinalizeBlock,
+					) (haltchain bool) {
+						// Note the first TX is MsgProposeOperations, the second is all other TXs.
+						execResult := response.TxResults[1]
+						require.True(t, execResult.IsErr())
+						require.Equal(t, satypes.ErrFailedToUpdateSubaccounts.ABCICode(), execResult.Code)
```

Comment:
Planning to add the following additional tests:
- Transfers and withdrawals become unblocked after `WITHDRAWAL_AND_TRANSFERS_BLOCKED_AFTER_NEGATIVE_TNC_SUBACCOUNT_SEEN_BLOCKS` blocks
- Both transfers and withdrawals fail
- Transfers and withdrawals are not blocked if subaccounts specified in negative TNC subaccounts have non-negative TNC after liquidations and deleveraging steps